airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io/airbyte/cdk/read/cdc/CdcPartitionReader.kt

```diff
@@ -205,6 +229,35 @@ class CdcPartitionReader<T : Comparable<T>>(
         )
     }
 
+    /**
+     * Watchdog coroutine that monitors for timeout when no events are received from Debezium. This
+     * is necessary because Debezium may not emit any events (including heartbeats) when the
+     * database has no changes, causing the sync to hang indefinitely.
+     *
+     * The watchdog waits for exactly the timeout duration, then checks once. If any event is
+     * received during that time, the watchdog is cancelled by the event consumer.
+     */
+    private suspend fun startWatchdog(timeoutDuration: Duration) {
+        log.info { "Starting watchdog with timeout of ${timeoutDuration.seconds} seconds" }
+        delay(timeoutDuration.toMillis())
+
+        // If we reach here, no event was received within the timeout period
+        // Check if we should still proceed (engine might have closed for other reasons)
+        if (watchdogShouldStop || closeReasonReference.get() != null) {
+            log.info { "Watchdog woke up but engine already stopped" }
+            return
+        }
+
+        log.info {
+            "Watchdog timeout: no events received for ${timeoutDuration.toMinutes()} minutes. " +
+                "Records emitted: ${numEmittedRecords.get()}. Shutting down Debezium engine."
+        }
+
+        if (closeReasonReference.compareAndSet(null, CloseReason.WATCHDOG_TIMEOUT)) {
+            runBlocking { launch(Dispatchers.IO + Job()) { engine.close() } }
```

Comment:
Just making sure: 
Debezium engine will get stuck of it's getting closed from the same thread it is running on. I don't think that's the case here.